File: packages/reference/src/parse.ts
/**
 * Parsing helpers for arc.status events.
 *
 * These wrap `JSON.parse` with structural validation. They never throw on
 * malformed input; they return a `ValidationResult`.
 */

import { validate, type ValidationResult } from "./validate.js";

/**
 * Parse a single JSON string or buffer into a validated event.
 */
export function parse(input: string | Buffer): ValidationResult {
  const text = typeof input === "string" ? input : input.toString("utf8");

  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (err) {
    return {
      ok: false,
      issues: [
        {
          path: "",
          message: `not valid JSON: ${(err as Error).message}`,
        },
      ],
    };
  }

  return validate(parsed);
}

export interface JsonlParseResult {
  /** Validated events in order of appearance. */
  events: import("./types.js").ArcStatusEvent[];
  /** Per-line errors keyed by 0-indexed line number. */
  errors: Array<{
    line: number;
    issues: import("./validate.js").ValidationIssue[];
  }>;
}

/**
 * Parse a JSON Lines stream of events. Empty lines are skipped.
 *
 * Does not throw. Lines that fail to parse or validate are reported in
 * `errors`; valid lines are returned in `events`. Callers decide whether
 * to treat any error as fatal.
 */
export function parseJsonl(input: string | Buffer): JsonlParseResult {
  const text = typeof input === "string" ? input : input.toString("utf8");
  const lines = text.split(/\r?\n/);

  const events: import("./types.js").ArcStatusEvent[] = [];
  const errors: JsonlParseResult["errors"] = [];

  for (let i = 0; i < lines.length; i++) {
    const raw = lines[i];
    if (raw === undefined || raw.trim() === "") continue;
    const result = parse(raw);
    if (result.ok) {
      events.push(result.event);
    } else {
      errors.push({ line: i, issues: result.issues });
    }
  }

  return { events, errors };
}

File: reference/python/src/agent_arc_status/parse.py
"""Parsing helpers: wrap ``json.loads`` with validation. Never raise."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass

from .types import ArcStatusEvent
from .validate import ValidationErr, ValidationIssue, ValidationOk, ValidationResult, validate

# Mirror the TS ``split(/\r?\n/)`` exactly; ``str.splitlines()`` also breaks on
# \v, \f, \x1c and U+2028 which would diverge on adversarial input.
_LINE_SPLIT = re.compile(r"\r?\n")


def parse(data: str | bytes) -> ValidationResult:
    """Parse a single JSON string/bytes into a validated event."""
    text = data.decode("utf-8") if isinstance(data, bytes) else data
    try:
        parsed = json.loads(text)
    except ValueError as err:
        return ValidationErr([ValidationIssue("", f"not valid JSON: {err}")])
    return validate(parsed)


@dataclass(frozen=True, slots=True)
class JsonlLineError:
    line: int
    issues: list[ValidationIssue]


@dataclass(frozen=True, slots=True)
class JsonlParseResult:
    events: list[ArcStatusEvent]
    errors: list[JsonlLineError]


def parse_jsonl(data: str | bytes) -> JsonlParseResult:
    """Parse a JSON Lines stream. Empty lines are skipped; errors are collected
    per 0-indexed line rather than raised."""
    text = data.decode("utf-8") if isinstance(data, bytes) else data
    lines = _LINE_SPLIT.split(text)

    events: list[ArcStatusEvent] = []
    errors: list[JsonlLineError] = []

    for i, raw in enumerate(lines):
        if raw.strip() == "":
            continue
        result = parse(raw)
        if isinstance(result, ValidationOk):
            events.append(result.event)
        else:
            errors.append(JsonlLineError(i, result.issues))

    return JsonlParseResult(events, errors)
